File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_60_reverse_node_in_k_group.ts
import { ListNode } from './ListNode';   

function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    // Check if we have k nodes to reverse
    let curr = head;
    let count = 0;
    while (curr && count < k) {
        curr = curr.next;
        count++;
    }
    
    // If we have k nodes, reverse them
    if (count === k) {
        // Reverse k nodes
        let prev: ListNode | null = null;
        curr = head;
        for (let i = 0; i < k; i++) {
            const nextNode = curr!.next;
            curr!.next = prev;
            prev = curr;
            curr = nextNode;
        }
        
        // head is now the last node in reversed group
        // prev is the new head of this group
        // curr is the start of next group
        // Recursively reverse next groups
        head!.next = reverseKGroup(curr, k);
        return prev;
    }
    
    // Less than k nodes remaining, return as is
    return head;
}     

File: src/my_project/interviews/top_150_questions_round_22/ex_60_reverse_node_in_k_group.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # Check if we have k nodes to reverse
        curr = head
        count = 0
        while curr and count < k:
            curr = curr.next
            count += 1
        
        # If we have k nodes, reverse them
        if count == k:
            # Reverse k nodes
            prev = None
            curr = head
            for _ in range(k):
                next_node = curr.next
                curr.next = prev
                prev = curr
                curr = next_node
            
            # head is now the last node in reversed group
            # prev is the new head of this group
            # curr is the start of next group
            # Recursively reverse next groups
            head.next = self.reverseKGroup(curr, k)
            return prev
        
        # Less than k nodes remaining, return as is
        return head